packages/openapi-fetch/src/index.js

```diff
@@ -144,77 +162,224 @@ export default function createClient(clientOptions) {
 // utils
 
 /**
- * Serialize query params to string
- * @type {import("./index.js").defaultQuerySerializer}
+ * Serialize primitive param values
+ * @type {import("./index.js").serializePrimitiveParam}
  */
-export function defaultQuerySerializer(q) {
-  const search = [];
-  if (q && typeof q === "object") {
-    for (const [k, v] of Object.entries(q)) {
-      const value = defaultQueryParamSerializer([k], v);
-      if (value) {
-        search.push(value);
-      }
-    }
+export function serializePrimitiveParam(name, value, options) {
+  if (value === undefined || value === null) {
+    return "";
+  }
+  if (typeof value === "object") {
+    throw new Error(
+      `Deeply-nested arrays/objects aren’t supported. Provide your own \`querySerializer()\` to handle these.`,
+    );
   }
-  return search.join("&");
+  return `${name}=${options?.allowReserved === true ? value : encodeURIComponent(value)}`;
 }
 
 /**
- * Serialize query param schema types according to expected default OpenAPI 3.x behavior
- * @type {import("./index.js").defaultQueryParamSerializer}
+ * Serialize object param (shallow only)
+ * @type {import("./index.js").serializeObjectParam}
  */
-export function defaultQueryParamSerializer(key, value) {
-  if (value === null || value === undefined) {
-    return undefined;
+export function serializeObjectParam(name, value, options) {
+  if (!value || typeof value !== "object") {
+    return "";
   }
-  if (typeof value === "string") {
-    return `${deepObjectPath(key)}=${encodeURIComponent(value)}`;
+  const values = [];
+  const joiner =
+    {
+      simple: ",",
+      label: ".",
+      matrix: ";",
+    }[options.style] || "&";
+
+  // explode: false
+  if (options.style !== "deepObject" && options.explode === false) {
+    for (const k in value) {
+      values.push(
+        k,
+        options.allowReserved === true
+          ? value[k]
+          : encodeURIComponent(value[k]),
+      );
+    }
+    const final = values.join(","); // note: values are always joined by comma in explode: false (but joiner can prefix)
+    switch (options.style) {
+      case "form": {
+        return `${name}=${final}`;
+      }
+      case "label": {
+        return `.${final}`;
+      }
+      case "matrix": {
+        return `;${name}=${final}`;
+      }
+      default: {
+        return final;
+      }
+    }
   }
-  if (typeof value === "number" || typeof value === "boolean") {
-    return `${deepObjectPath(key)}=${String(value)}`;
+
+  // explode: true
+  for (const k in value) {
```

Comment:
In the vein of performance, both `Object.entries()` and `for … in` are (really) fast. But `for … in` is about 2× faster. And though this is totally splitting hairs, I’d be remiss not to use whatever is an order of magnitude faster if there’s no discernable readability/complexity impact.

This was tested in current V8, and this smells like a temporary thing that will flatten out so both will probably be equally performant over time. But I saw the numbers today, and had to do it 🤷